client/src/panel-cnxml-preview.ts

```diff
@@ -1,105 +1,230 @@
-    return [contents, editor, activeDocument.uri]
-  }
-  return [contents, editor, uri]
+export type PanelIncomingMessage = (
+  DirectEditIncoming
+  | ScrollInEditorIncoming
+)
+
+export interface RefreshOutgoing {
+  type: 'refresh'
+  xml: string
 }
 
-export const showCnxmlPreview = (panelType: PanelType, resourceRootDir: string, activePanelsByType: {[key in PanelType]?: vscode.WebviewPanel}) => async () => {
-  let [contents, editor, resource] = getContents()
-  if (contents == null || resource == null) { return }
-  const definitelyResource = resource
+// Line is one-indexed
+export interface ScrollToLineOutgoing {
+  type: 'scroll-in-preview'
+  line: number
+}
 
-  const resourceColumn = editor?.viewColumn ?? vscode.ViewColumn.One
-  const previewColumn = resourceColumn + 1 // because the preview is on the side
+export type PanelOutgoingMessage = (
+  RefreshOutgoing
+  | ScrollToLineOutgoing
+)
 
+const ELEMENT_NODE = 1
+export const tagElementsWithLineNumbers = (doc: Document): void => {
+  const root = doc.documentElement
+  const stack: Element[] = [root]
+  while (stack.length > 0) {
+    const current = expect(stack.pop(), 'stack length is non-zero')
+    current.setAttribute('data-line', (current as any).lineNumber)
+    stack.push(...Array.from(current.childNodes).filter(node => node.nodeType === ELEMENT_NODE) as Element[])
+  }
+}
+
+const initPanel = (context: ExtensionHostContext) => () => {
+  const editor = vscode.window.activeTextEditor
+  const resourceColumn = editor?.viewColumn ?? vscode.ViewColumn.One
+  const previewColumn = resourceColumn + 1
+  const localResourceRoots = [vscode.Uri.file(context.resourceRootDir), expect(getRootPathUri(), 'workspace must be open to preview')]
   const panel = vscode.window.createWebviewPanel(
-    panelType,
-    `Preview ${path.basename(resource.fsPath)}`,
+    PanelType.CNXML_PREVIEW,
+    'CNXML Preview',
     previewColumn, {
       enableScripts: true,
-      localResourceRoots: getLocalResourceRoots([vscode.Uri.file(resourceRootDir)], resource),
+      localResourceRoots,
       enableFindWidget: true
     }
   )
-  activePanelsByType[panelType] = panel
-  let disposed = false
-
-  let html = fs.readFileSync(path.join(resourceRootDir, 'cnxml-preview.html'), 'utf-8')
-  html = addBaseHref(panel.webview, resource, html)
-  html = fixResourceReferences(panel.webview, html, resourceRootDir)
-  html = fixCspSourceReferences(panel.webview, html)
-  panel.webview.html = html
-
-  const xml = contents
-  await panel.webview.postMessage({ xml })
-  let throttleTimer = setTimeout(() => {
-    updatePreview().catch((err) => { throw new Error(err) })
-  }, 200)
-
-  async function updatePreview(): Promise<void> {
-    clearTimeout(throttleTimer)
-    if (disposed) {
+  panel.webview.html = rawTextHtml('Loading...')
+  return panel
+}
+
+export const isCnxmlFile = (document: vscode.TextDocument): boolean => {
+  return document.languageId === 'cnxml'
```

Comment:
Does this ever return `true`? I believe due to the following in our `settings.json`, for CNXML `languageId` will be `xml`:

```json
"files.associations": {
    "*.cnxml": "xml"
}
```

This wouldn't impact correctness where `isCnxmlFile` is used today since `isXmlFile` will return `true` but  could possibly cause confusion in the future 🤷‍♂️ . 

Just as a note: The association of `*.cnxml` with `xml` in `settings.json` goes back to [this](https://github.com/openstax/cnx/issues/1354#issuecomment-772862496).